File: Medium/41- 155. Min Stack/array-solution.go
type entry struct {
	val int
	min int
}

type MinStack struct {
	stack []entry
}

func Constructor() MinStack {
	return MinStack{}
}

func (s *MinStack) Push(value int) {
	m := value
	if !s.IsEmpty() {
		m = min(value, s.GetMin())
	}
	s.stack = append(s.stack, entry{val: value, min: m})
}

func (s *MinStack) Pop() {
	s.stack = s.stack[:len(s.stack)-1]
}

func (s *MinStack) Top() int {
	return s.stack[len(s.stack)-1].val
}

func (s *MinStack) GetMin() int {
	return s.stack[len(s.stack)-1].min
}

func (s *MinStack) IsEmpty() bool {
	return len(s.stack) == 0
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(value);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: Medium/41- 155. Min Stack/array-solution.ts
class MinStack {
    private arr: Array<{ val: number, min: number }>
    constructor() {
        this.arr = []
    }
    
    push(value: number): void {
        this.arr.push({
            val :value,
            // this works because every inserted number knows it's the min or not
            // and if we poped the smallest number, the last one before it would 
            // still have the second smallest min
            min: this.isEmpty()? value: Math.min(value, this.getMin())
        })
    }

    pop(): void {
        this.arr.pop()
    }

    top(): number {
        return this.arr[this.arr.length - 1].val
    }

    getMin(): number {
        return this.arr[this.arr.length - 1].min
    }

    isEmpty(): boolean {
        return this.arr.length === 0
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(value)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
